### ISSUE TITLE
Save and verify hashes when locking dependencies

### ISSUE THREAD

**Issue by ligi** [gradle/gradle#5633]
I was really excited when reading the release notes for Gradle 4.8! Especially the part about "dependency locking". Unfortunately then I saw it does not yet fully do what I expected it to do. I thought it might be a replacement for [gradle-witness](https://github.com/signalapp/gradle-witness) - which would be really awesome and solve some pain-points I have currently. Unfortunately there is a small thing missing so it can replace gradle-witness fully. If the hashes of the artifacts would be also saved in the lockfile and checked - then I could drop gradle-witness (or better the fork by @akwizgran I am using currently that can deal with api/implementation) 

Hence this feature request. Thanks for considering!

Context:
 * https://docs.gradle.org/4.8/release-notes.html
 * https://docs.gradle.org/4.8/userguide/dependency_locking.html

**Comment by oehme** [comment]
This sounds like a completely different use case from locking. Why not keep using the witness plugin for that verification part?

**Comment by ligi** [comment]
I think the use-cases are quite close together. Both want to lock down a version - gradle-witness just locks it down a bit more than the current implementation of dependency locking in gradle 4.8.
For me it would make life easier as currently updating the dependency block in the gradle file is a bit painful and it also clutters the gradle file a lot see: https://github.com/walleth/walleth/blob/c41141a3c69472f2c8419fc9fbf21b8e0e1bbe3c/app/build.gradle#L233
Also when things change in gradle then the plugin has to be adjusted - had a long time where gradle-witness was not working at all when using with new AS/gradle - in TREZOR-Android there is still a problem in one module - context: https://github.com/trezor/trezor-android/pull/34
Another thing which is a bit annoying with the plugin solution currently is that having to add a jar to every repo is not optimal - would be really nice if this would be part of gradle - also knows some devs don't want to use this from a plugin but expect it from the build-system - e.g @schildbach

**Comment by oehme** [comment]
I don't understand how automatically adding checksums would add any security. When you generate the lock, you've already resolved. So you might already have downloaded an untrusted file and locked down the untrusted checksum.

**Comment by ligi** [comment]
But you check-in the lock file to version control and I always build the apk to upload on the CI server. Sure it does not guarantee full security - but it is one more barrier a potential attacker has to overcome. Also on top of this you can build some more verifications. E.g. multiple CI servers that execute a task with `gradle dependencies --write-locks.` and if there is a change in checksums - then fail ..

**Comment by yogurtearl** [comment]
> and if there is a change in checksums - then fail ..

That is important not only for security, but also for reproducible builds. I want to know that it is using the exact same bits when if I rebuild it 6 months from now.

**Comment by melix** [comment]
The two problems are related but orthogonal: you may want to check checksums without locking, and you may want locking without checksums. Gradle already saves the dependencies in its local repository with enough information, it's just not available. In particular, the hashes of artifacts are known.

Note also that for a single dependency you can have more than one artifact, so a single signature is not enough. Also, you might want to check the SHA1 or ASC signatures. So I think we should keep the 2 aspects distinct.

In a nutshell, dependency locking assumes that the dependencies are _not changing_, which means, for a single version string, the artifacts are _always the same_. Then, you may want additional guarantees on top, but it's a different aspect: for a single version string, *and for the artifacts that were downloaded for this specific resolution*, a checksum for each of them.

**Comment by ligi** [comment]
> you may want to check checksums without locking

what use case do you see for this

**Comment by vlsi** [comment]
>>you may want to check checksums without locking

>what use case do you see for this

For instance: "I trust official Apache Commons releases, however I don't really care which version of commons-io.jar is used"
The trust can be expressed as "I trust official Apache Commons" can be expressed "all artifact files must be signed by at least one of the keys from https://www.apache.org/dist/commons/KEYS"

It would be nice if Gradle could download ASC files automatically and expose those along with jar file for the inspection.

PS. I've created https://github.com/vlsi/vlsi-release-plugins/commit/208734b6e9d67c11d0d697f883f3e23107284820 that verifies checksums based on [DependencyResolutionListener](https://github.com/gradle/gradle/blob/v5.5.1/subprojects/core-api/src/main/java/org/gradle/api/artifacts/DependencyResolutionListener.java)

**Comment by melix** [comment]
Thanks. We plan on working on checksum and signature checking this fall.

**Comment by vlsi** [comment]
Just in case, here's a random datapoint from Apache JMeter build: https://travis-ci.org/apache/jmeter/jobs/566878604#L1257

`System.currentTimeMillis` shows that Travis was able to compute 164 SHA-512 sums in 1.4sec over 104'775'785 bytes of data.
This gives roughly 73MiB/sec.
PGP verification would probably take longer.

I guess the corollary is "it does not take ages" (~full verification at Travis is ok), and "Gradle would like to cache the results of checksum verification" (to avoid 1-2-3 second delays for developer's `./gradlew` experience).

**Comment by vlsi** [comment]
PS. I've implemented PGP verification in https://github.com/vlsi/vlsi-release-plugins/tree/master/plugins/checksum-dependency-plugin

**Comment by rgoldberg** [comment]
There seem to be three related checksum / signature / locking features being discussed (from now on, I'll just mention checksums to be concise):

1. Download verification: comparing a checksum computed from a downloaded artifact against a checksum supplied by a server, to attempt to ensure that the artifact contains what the server says it should contain.

2. Dependency content locking: comparing a checksum computed from a downloaded artifact against a checksum stored in a lock file, to ensure that the artifact used by the current build is the same as was used in previous build.

3. Dependency version locking: comparing a resolved dependency version against a version stored in a lock file.  This is already implemented in Gradle 4.8.

@melix mentioned that the Gradle team plans on "working on checksum and signature checking this fall".  Is that download verification, dependency content locking, or both?

As long as checksums are supplied by the server, I don't see much advantage in skipping download verification, except for potential performance reasons.

I guess that there are uses for download verification without dependency content locking and/or dependency version locking (though I'd always want to lock), but I don't see much advantage in locking dependency versions without content (except for performance), or for locking dependency content without versions (a version check should be fast, and, in the unlikely event that a version changes without a content change, it wouldn't be hard to update the version).

Would saving the file size in addition to the checksum be useful for dependency content locking? If the file sizes are different, the lock would fail without needing to calculate a checksum.

**Comment by vlsi** [comment]
@rgoldberg , you seem to miss "integrity verification".

For instance:

>1. comparing a checksum computed from a downloaded artifact against a checksum supplied by a server

Suppose a developer declares `"commons-lang:commons-lang"` dependency.
A malicious server could supply a malware instead of a proper `commons-lang.jar`. Of course it can supply a valid SHA checksum for that malware as well.

That is why build configuration should specify expected signatures and/or hashes.

**Comment by rgoldberg** [comment]
@vlsi That use case is covered by dependency checksum locking (i.e. comparing against a locked checksum)

**Comment by vlsi** [comment]
@rgoldberg , I'm afraid I fail to see a question and/or suggestion in your comments.

**Comment by ligi** [comment]
> Would saving the file size in addition to the checksum be useful for dependency content locking? If the file sizes are different, the lock would fail without needing to calculate a checksum.

I do not like the idea. It would add complexity without significant gain.

**Comment by rgoldberg** [comment]
@vlsi: I asked two questions in my initial comment.  Just search for `?`.

Beforehand, I delineated the different features being discussed so that I could accurately convey my first question, as the meaning of "checksum and signature checking" was ambiguous.

I also implied suggestions & questions about using the delineated features in combination, or separately, by discussing the reasons that I saw for using or not using them.  e.g., it seems pretty obvious to me that I suggested that Gradle probably needn't support dependency content locking without dependency version locking, but that it should support version locking without content locking because of SNAPSHOT releases (added in an edit), and possibly because of performance concerns.

I then asked about the usefulness of a suggested performance improvement to overcome the aforementioned potential performance issues.

My second comment suggested that your "integrity verification" use case would be handled via feature 2 delineated in my first comment, and contained instructions about how to do so.

My comments therefore seem to me to contain both questions & suggestions.

**Comment by vlsi** [comment]
> rgoldberg: mentioned that the Gradle team plans on "working on checksum and signature checking this fall". Is that download verification, dependency content locking, or both?

That was a question to melix (not mentioning because they probably get too much notifications already), however it is apparent that "signatures" are meant to be used for content verification. There's no much sense in using signatures for "download verification" purpose.

That is why I think this is not a question: the answer is obvious. Just in case, the answer is "both".

> rgoldberg: Would saving the file size in addition to the checksum be useful for dependency content locking?

This is out of scope of hash/signature verification. Apparently Gradle has well-established ways of tracking stale data (e.g. for tracking up-to-date task inputs-outputs), so I see no reason to discuss "optimizations". There's just no point in discussing file sizes, because the nature of hash/signature verification is still moot. It is not clear what should be configurable, and so on.

> it seems pretty obvious to me that I suggested that Gradle probably needn't support dependency content locking without dependency version locking

It seems obvious that you mention "dependency version locking" without making it clear if you consider `implementation("commons-lang:commons-lang:2.6")` to be "dependency version locking" or not.
Technically speaking, this is a basic dependency declaration, however you mention `locking` too much.

The feature of hash/signature verification is of a great use for basic dependency declarations with explicit versions.


Both sentences with ? sign turn out to classify as not-a-question.


> it seems pretty obvious to me that I suggested that Gradle probably needn't support dependency content locking without dependency version locking

It seems pretty obvious you miss the case for a trust-based PGP verification (see https://github.com/gradle/gradle/issues/5633#issuecomment-517603552 )
For instance, I want to specify that "I trust all `commons-lang` jars provided they are signed with a proper PGP key".
That enables to keep certain level of security (==use only official binaries in my builds), however I don't have to explicitly lock commons-lang version. I can use wildcards or whatever, and the requirement is the artifact has to be PGP-signed.

This use-case is a counter-example to your "needn't support dependency content locking without dependency version locking".

That is very popular use case for having PGP in dependency management at all.

> I then asked about the usefulness of a suggested performance improvement

There's nothing to discuss yet. I'm sure you know what premature optimization is.
Gradle-core has no publicly-available PGP verification, so there's nothing to discuss performance-wize.

On the other hand, my `com.github.vlsi.gradle:checksum-dependency-plugin` plugin is available, and it works. So you can try how it works on real-life projects rather than discuss theoretical "file size" suggestions.

It turns out the vital feature for PGP verification is ability to failover across multiple PGP keyservers. For some reason keyservers respond with 500, 504 and so on HTTP codes, so key downloader has to implement client-site download request balancing.
It is something I'm working on, and it will land to checksum-dependency shortly.

**Comment by rgoldberg** [comment]
@vlsi I haven’t yet read your whole screed, but if people like you bring up the performance of checksums & signatures, I would think that a possible performance shortcut is completely applicable. Don’t tell me that you can bring up a topic like performance and no one else can mention it. Also, don’t tell me that it’s obvious that

**Comment by vlsi** [comment]
> but if people like you bring up the performance of checksums & signatures

The difference is I have implemented a solution, so I measure real-life cases on a workable code, and your claims are purely theoretical.

> Don’t tell me that you can bring up a topic like performance and no one else can mention it

Go ahead. Implement a workable solution, then we would have a base ground.

As I said: 

> Vladimir: I guess the corollary is "**it does not take ages**" (~**full verification at Travis is ok**)

In other words, preliminary analysis shows that even a very naive implementation is just fine.

**Comment by ljacomet** [comment]
While it is great to see interest in this, and exchange of ideas, I will kindly ask that it remains polite and conflict free.

**Comment by rgoldberg** [comment]
@vlsi: I’m not trying to be argumentative, but I really just don’t understand your assertion:

> It seems obvious that you mention "dependency version locking" without making it clear if you consider implementation("commons-lang:commons-lang:2.6") to be "dependency version locking" or not.
> Technically speaking, this is a basic dependency declaration, however you mention locking too much.

In my first comment, I explicitly defined:

> Dependency version locking: comparing a resolved dependency version against a version stored in a lock file. This is already implemented in Gradle 4.8.

The dependency line that you mentioned is from a build file, not a lock file. It was also supported before Gradle 4.8, whereas lock file support was added in 4.8.

If you can explain why my definition confused you, I will happily edit my comment to make it clear for anyone else who might read it in the future. Should I change “lock file” to “`*.lockfile`“?

**Comment by vlsi** [comment]
@rgoldberg , I would kindly ask you to avoid confining this issue/feature to `lockfile` use-case only.

Verification of hash/signatures has lots of use cases even without `lockfile`.

**Comment by vlsi** [comment]
@ligi, would you please update issue title to accommodate not just `lockfile` use-cases, but overall use of hash/signatures for the verification of the used dependencies/plugins?

**Comment by rgoldberg** [comment]
@vlsi: Yet again, I don’t understand why you’re saying what you’re saying:

> @rgoldberg , I would kindly ask you to avoid confining this issue/feature to lockfile use-case only.

Can you please explain to me how I’ve confined this to lockfile use-case only?

In the first of the 3 features I mentioned, I discussed comparing two things that don’t involve lock files at all.

For the second two features, lock files seem to me to be a more appropriate place than `build.gradle[.kts]` files to provide the information to lock versions and/or content (via checksum and/or signature).

If I’ve written something that’s ambiguous, please:

1) help me refine my comments to be unambiguous, as I’ve already asked
2) don’t criticize my comments yet again without first helping me fix whatever potential ambiguities might exist to cause you to have such wildly different readings than I intended

**Comment by ligi** [comment]
@vlsi what title do you suggest?

**Comment by vlsi** [comment]
Naming is hard, however something like "Verify hashes and PGP signatures for dependencies and plugins" would probably do.

**Comment by rgoldberg** [comment]
@vlsi I don’t understand:

> It seems pretty obvious you miss the case for a trust-based PGP verification (see #5633 (comment) )

I thought that that use case was covered by my feature 1, which is comparing a downloaded checksum / signature with that calculated from the downloaded artifact.  If you think that that isn’t covered by my feature 1, please explain to me how it isn’t.

Obviously, if you are using signatures instead of checksums, you need a way to specify the applicable public keys, but nothing in any of the theee features precluded specifying public keys / using signatures.

**Comment by vlsi** [comment]
It is good you classify "feature 1", "feature 2", "feature 3".

> I thought that that use case was covered by my feature 1, which is comparing a downloaded checksum / signature with the one calculated from the associated downloaded artifact. If you think that that isn’t covered by my feature 1, please explain to me how it isn’t.

The "feature 1" is:

> ... to ensure that the artifact contains what the **server says** it should contain

On contrary, PGP verification is to ensure that the artifact contains what **build script says** it should contain.

In other words, your feature 1 seems to address "unexpected failures" (e.g. a cosmic that hits RAM and corrupts a downloaded file). It is really a good feature (I guess it is already implemented), and you are right it is required in a modern build system.
However, I would suggest we don't discuss that feature here. It is not like the feature is bad.
I suggest we discuss security issues here like Gradle's ability to detect and prevent man-in-the-middle attacks.

**Comment by ligi** [comment]
> On contrary, PGP verification is to ensure that the artifact contains what build script says it should contain.

no - PGP verification cannot do this. You will need reproducible builds and multiple build servers for this. But this is a whole other can of worms. Would love to open this can at some point - but not now. I think we should go step by step. And the first step would be what I asked for in the issue.
Also a reason I decided not to change the title - I would love to not side-track this issue and stay focused.

**Comment by rgoldberg** [comment]
@vlsi In my initial comment, I only spoke about checksums for brevity. 

Given that checksums are generally provided by the same server, and that there’s nothing like a public key to check against, you have to go against what the server itself says.

For signatures, I had just assumed that everyone would modify what I said for checksums to be correct for signatures, which would definitely involve using a public key, and not just taking the server’s word for it.

I can see that my brevity was too terse, so I’ve added a concluding note to my feature 1 to indicate the different capabilities provided by signatures.

I chose to discuss checksums instead of signatures to align my comment as closely as possible to the OP’s initial issue description.

Sorry for any confusion.

**Comment by vlsi** [comment]
Please accept my apologies as I was wrong with my assumption that this issue includes "trust verification".
I do not care on "dependency locking" feature, so I have nothing to add here.

I've created https://github.com/gradle/gradle/issues/10443 to track "trust verification" and "man-in-the-middle prevention".

**Comment by melix** [comment]
I'm starting work on this larger topic of security in 6.+. @vlsi I took a look at your plugin and I must say this is almost exactly what I had in mind. I'll dive a bit more.

**Comment by melix** [comment]
It's extremely unlikely we'll ever bound dependency locking with checksum verification. Those are really orthogonal things.

I think the right approach is the one described in #10443, where _verification_ implies verification of checksums and/or signatures.

Dependency locking only "talks" about versions, nothing more.
Then, you can _in addition_ enable verification, just like you can use _verification_ without locking.

**Comment by melix** [comment]
Dependency verification is now in `master` and will be shipped in Gradle 6.2. As explained in the previous comment, this feature is orthogonal to dependency locking and can be used alongside locking.
